src/services_models/services_constructor.jl

```diff
@@ -41,6 +32,54 @@ function get_incompatible_devices(devices_template::Dict)
     return incompatible_device_types
 end
 
+function sort_contributing_devices(contributing_devices)
+    types = unique(typeof.(contributing_devices))
+    device_list = Dict()
+    for t in types
+        _devices = filter(x -> typeof(x) == t, contributing_devices)
+        device_list[t] = _devices
+    end
+    return device_list
+end
+
+function populate_contributing_devices!(
```

Comment:
this function might need some additional care. I don't fully understand what it is doing.